apiserver/common/imagecommon/imagemetadata_test.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/apiserver/common/imagecommon"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/state/cloudimagemetadata"
+	coretesting "github.com/juju/juju/testing"
+)
+
+type imageMetadataSuite struct {
+	st *mockState
+}
+
+var _ = gc.Suite(&imageMetadataSuite{})
+
+func (s *imageMetadataSuite) SetUpTest(c *gc.C) {
+	mCfg := testConfig(c)
+
+	s.st = &mockState{
+		Stub: testing.Stub{},
+		saveMetadata: func(m []cloudimagemetadata.Metadata) error {
+			return nil
+		},
+		modelCfg: func() (*config.Config, error) {
+			return mCfg, nil
+		},
+	}
+}
+
+func (s *imageMetadataSuite) TestSaveEmpty(c *gc.C) {
+	errs, err := imagecommon.Save(s.st, params.MetadataSaveParams{})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(errs, gc.HasLen, 0)
+	s.st.CheckCallNames(c, []string{}...)
+}
+
+func (s *imageMetadataSuite) TestSaveModelCfgFailed(c *gc.C) {
+	m := params.CloudImageMetadata{
+		Source: "custom",
+	}
+	msg := "save error"
+
+	s.st.modelCfg = func() (*config.Config, error) {
+		return nil, errors.New(msg)
+	}
+
+	errs, err := imagecommon.Save(s.st, params.MetadataSaveParams{
+		Metadata: []params.CloudImageMetadataList{{
+			Metadata: []params.CloudImageMetadata{m},
+		}},
+	})
+	c.Assert(errors.Cause(err), gc.ErrorMatches, msg)
+	c.Assert(errs, gc.IsNil)
+	s.st.CheckCallNames(c, "ModelConfig")
+}
+
+func (s *imageMetadataSuite) TestSave(c *gc.C) {
+	m := params.CloudImageMetadata{
+		Source: "custom",
+	}
+	msg := "save error"
+
+	saveCalls := 0
+	s.st.saveMetadata = func(m []cloudimagemetadata.Metadata) error {
+		saveCalls += 1
+		c.Assert(m, gc.HasLen, saveCalls)
+		// TODO (anastasiamac 2016-08-24) This is a check for a band-aid solution.
+		// Once correct value is read from simplestreams, this needs to go.
+		// Bug# 1616295
+		// Ensure empty stream is changed to release
+		c.Assert(m[0].Stream, gc.DeepEquals, "released")
+		if saveCalls == 1 {
+			// don't err on first call
+			return nil
+		}
+		return errors.New(msg)
+	}
+
+	errs, err := imagecommon.Save(s.st, params.MetadataSaveParams{
+		Metadata: []params.CloudImageMetadataList{{
+			Metadata: []params.CloudImageMetadata{m},
+		}, {
+			Metadata: []params.CloudImageMetadata{m, m},
+		}},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(errs, gc.HasLen, 2)
+	c.Assert(errs[0].Error, gc.IsNil)
+	c.Assert(errs[1].Error, gc.ErrorMatches, msg)
+	s.st.CheckCallNames(c, "ModelConfig", "SaveMetadata", "SaveMetadata")
+}
+
+type mockState struct {
+	testing.Stub
+	saveMetadata func(m []cloudimagemetadata.Metadata) error
+	modelCfg     func() (*config.Config, error)
+}
+
+func (s *mockState) SaveMetadata(m []cloudimagemetadata.Metadata) error {
+	s.MethodCall(s, "SaveMetadata", m)
+	return s.saveMetadata(m)
+}
+
+func (s *mockState) ModelConfig() (*config.Config, error) {
+	s.MethodCall(s, "ModelConfig")
+	return s.modelCfg()
```

Comment:
If you were to store the result of `testConfig()` as `mockState.mcfg`, you could make this:

```go
return s.mcfg, s.st.NextErr()
```

then drop s.saveMetadata and s.modelCfg